File: python/reverseInteger.py
# Time -> 0(N) Space -> 0(N)
def reverse(x: int) -> int:
    val = str(x)
    start = 0
    if val[0] == "-":
        start = 1
    end = len(val) - 1

    valStr = [s for s in val]

    while start < end:
        valStr[start], valStr[end] = valStr[end], valStr[start]
        start += 1
        end -= 1

    reversed = "".join(valStr)
    valInt = int(reversed)
    if valInt < pow(-2, 31) or valInt > pow(2, 31) - 1:
        return 0

    return valInt


print(reverse(-456))

File: go/ReverseInteger/ReverseInteger.go
package main

import (
	"fmt"
	"math"
	"strconv"
)

// Time -> 0(N) Space -> 0(N)
func reverse(x int) int {
	val := fmt.Sprintf("%d", x)

	runes := []rune(val)
	start := 0
	if runes[0] == rune('-') {
		start = 1
	}

	for i, j := start, len(runes)-1; i < j; i, j = i+1, j-1 {
		runes[i], runes[j] = runes[j], runes[i]
	}

	val = string(runes)
	valInt, _ := strconv.ParseInt(val, 10, 64)

	if valInt < int64(math.Pow(float64(-2), float64(31))) || valInt > int64(math.Pow(float64(2), float64(31)))-1 {
		return 0
	}

	return int(valInt)
}

func main() {
	fmt.Println(reverse(-456))
}
